plugins/org.jkiss.dbeaver.ext.postgresql/tests/org/jkiss/dbeaver/ext/postgresql/model/PostgreWithClauseAppenderTest.java

```diff
@@ -0,0 +1,127 @@
+        table = mock(GreenplumTable.class);
+
+        when(serverExtension.supportsOids()).thenReturn(false);
+        when(dataSource.getServerType()).thenReturn(serverExtension);
+        when(table.getDataSource()).thenReturn(dataSource);
+        when(table.isHasOids()).thenReturn(false);
+
+        when(tableBase.getRelOptions()).thenReturn(new String[]{"appendonly=true"});
+
+        String withClause = generateWithClause(table, tableBase);
+        assertEquals("\nWITH (\n\tappendonly=true\n)", withClause);
+    }
+
+    @Test
+    public void generateWithClause_whenTableIsGreenPlumTableWithOidsAndOptions_shouldDisplayWithClauseWithOidsAndOptions() {
+        table = mock(GreenplumTable.class);
+
+        when(serverExtension.supportsOids()).thenReturn(true);
+        when(dataSource.getServerType()).thenReturn(serverExtension);
+        when(table.getDataSource()).thenReturn(dataSource);
+        when(table.isHasOids()).thenReturn(true);
+
+        when(tableBase.getRelOptions()).thenReturn(new String[]{"appendonly=true"});
+
+        String withClause = generateWithClause(table, tableBase);
+        assertEquals("\nWITH (\n\tOIDS=TRUE, appendonly=true\n)", withClause);
+    }
+
+    @Test
+    public void generateWithClause_whenTableIsGreenPlumTableWithoutOidsAndWithoutOptions_shouldNotDisplayWithClause() {
+        table = mock(GreenplumTable.class);
+
+        when(serverExtension.supportsOids()).thenReturn(false);
+        when(dataSource.getServerType()).thenReturn(serverExtension);
+        when(table.getDataSource()).thenReturn(dataSource);
+        when(table.isHasOids()).thenReturn(false);
+
+        when(tableBase.getRelOptions()).thenReturn(null);
+
+        String withClause = generateWithClause(table, tableBase);
+        assertEquals("", withClause);
+    }
+
+    @Test
+    public void generateWithClause_whenTableIsGreenPlumTableWithOidsWithoutOptions_shouldDisplayWithClauseWithOids() {
+        table = mock(GreenplumTable.class);
+
+        when(serverExtension.supportsOids()).thenReturn(true);
+        when(dataSource.getServerType()).thenReturn(serverExtension);
+        when(table.getDataSource()).thenReturn(dataSource);
+        when(table.isHasOids()).thenReturn(true);
+
+        when(tableBase.getRelOptions()).thenReturn(null);
+
+        String withClause = generateWithClause(table, tableBase);
+        assertEquals("\nWITH (\n\tOIDS=TRUE\n)", withClause);
+    }
+
+    @Test
+    public void generateWithClause_whenTableIsGreenPlumTableWithOidsWithMultipleOptions_shouldDisplayWithClauseWithOidsAndAllTheOptions() {
+        table = mock(GreenplumTable.class);
+
+        when(serverExtension.supportsOids()).thenReturn(true);
+        when(dataSource.getServerType()).thenReturn(serverExtension);
+        when(table.getDataSource()).thenReturn(dataSource);
+        when(table.isHasOids()).thenReturn(true);
```

Comment:
Looks like each test has similar code blocks. Why don't you extract this code blocks to setup method.

Reply:
Refactored the test file to move this setup to a private setup helper method. The logic can't cleanly go into a '@Before' method as the when mocks for supportOids and isHasOids do change. Now each test calls setupGeneralWhenMocks and pass two booleans to indicate how it wants these to when mocks setup.

The alternative is to have the '@Before' set default when mocks and then have those tests that want different behavior override the default(s) by setting their own when.